### ISSUE TITLE
[Pipelines] Support for T2I-Adapter

### ISSUE THREAD

**Issue by wfng92** [huggingface/diffusers#2390]
### Model/Pipeline/Scheduler description

From the official repository, [T2I-Adapter](https://github.com/TencentARC/T2I-Adapter) by @TencentARC is

> ... a simple and small (~70M parameters, ~300M storage space) network that can provide extra guidance to pre-trained text-to-image models while freezing the original large text-to-image models.
> T2I-Adapter aligns internal knowledge in T2I models with external control signals. We can train various adapters according to different conditions, and achieve rich control and editing effects.

Would be great to have this plug and play adapters in `diffusers` module.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

Original code: https://github.com/TencentARC/T2I-Adapter
Pre-trained models: https://huggingface.co/TencentARC/T2I-Adapter
- [t2iadapter_keypose_sd14v1.pth](https://huggingface.co/TencentARC/T2I-Adapter/blob/main/models/t2iadapter_keypose_sd14v1.pth)
- [t2iadapter_seg_sd14v1.pth](https://huggingface.co/TencentARC/T2I-Adapter/blob/main/models/t2iadapter_seg_sd14v1.pth)
- [t2iadapter_sketch_sd14v1.pth](https://huggingface.co/TencentARC/T2I-Adapter/blob/main/models/t2iadapter_sketch_sd14v1.pth)

**Comment by HimariO** [comment]
I have a quick attempt to implement T2I-Adapter in diffusers over [here](https://github.com/HimariO/diffusers-t2i-adapter/blob/adapter/main.py)
This implementation is seen to be working correctly according to the result I got from the `t2iadapter_seg_sd14v1.pth` adapter:
![Screenshot from 2023-02-19 15-09-51](https://user-images.githubusercontent.com/12119935/219934245-a50ade58-10fb-4a98-9704-717257338fa9.png)

The adapter module itself is quite simple, so I think the main consideration of integrating the adapter into diffusers will be:
- Should the design of the adapter module allow it to inject the adapter hidden state to any layer inside UNet (in the official implementation the adapter state is always added after the last ResnetBlock from each downsample block)
- Should the adapter be integrated into UNet, since the number of feature maps and size of feature maps the adapter output all depend on the UNet model it is working with

**Comment by sayakpaul** [comment]
Thanks so much for your hard work, @HimariO! Your questions are quite valid. Do you think the design philosophy of how integrated LoRA into `diffusers` would be of help ([PR](https://github.com/huggingface/diffusers/pull/1884))? I mentioned it because LoRA also falls under the adapter series of neural nets.

Let me see what other members think.

Cc: @patil-suraj @patrickvonplaten @williamberman

**Comment by HimariO** [comment]
@sayakpaul, thank you for directing me to the LoRA PR. It has been very helpful in giving me an idea of the design philosophy of similar features. After reviewing the PR for LoRA and the draft PR for ControlNet, I believe we can create a more versatile API that can support T2I-Adapter, ControlNet, and other similar modules that have independent input and will inject the output into diffusion model. POC can be found [here](https://github.com/HimariO/diffusers-t2i-adapter/commit/3f2922110c4934fec9d5c240e346bd519a8c1948).

**Comment by haofanwang** [comment]
Agree. As T2I and ControlNet (they share similar design) both requires some change of UNet, more similar pipelines in the future may lead to crash. It is necessary to consider how to efficiently merge them into one framework.

**Comment by sayakpaul** [comment]
@HimariO I left a [comment](https://github.com/HimariO/diffusers-t2i-adapter/commit/3f2922110c4934fec9d5c240e346bd519a8c1948#commitcomment-101565530) directly on your commit. Thanks so much!

We usually consider the code-level impact we might have before accommodating a large change in the API. So, I request @patil-suraj @williamberman @patrickvonplaten @yiyixuxu to chime in here too. 

Note that this is a lighter-than-usual week for us, so there might be some delay in our response.

**Comment by takuma104** [comment]
My PR for ControlNet (#2407) has been open for a while now. I am also in favor of the Sideload-related changes. The T2I-Adapter and ControlNet share a similar concept in that they both interfere with UNet. I think that the Sideload concept could be a common foundation and have good potential for future extensions. (I have [left a comment](https://github.com/huggingface/diffusers/issues/2331#issuecomment-1440350516) on the ControlNet thread.)

**Comment by williamberman** [comment]
My understanding from a preliminary read of the t2i adapter paper is that the outputs from the adapter model are just added with the intermediate features of the encoder of the unet. This shouldn't require any hacking of the existing block definitions and could be done just by passing the outputs of the adapter to the forward method of the unet.

**Comment by HimariO** [comment]
@williamberman your understanding is correct, and what you describe is exactly what I do with my first prototype, The main motivation for trying out new concepts like sideloading is to avoid modifying every sub-module the adapter/controlnet-like model interacts with, especially when those modules are buried deep in the hierarchy.

**Comment by sayakpaul** [comment]
Thanks for thinking this through @HimariO! Let us know whenever you're read with a PR and / or if you need any help.

**Comment by AK391** [comment]
related: https://github.com/cloneofsimo/t2i-adapter-diffusers

**Comment by HimariO** [comment]
Hi @sayakpaul, just a quick note to let you know that I'm planning on creating the PR this week, and I'll let you know if there are any design-related issues that require further discussion. Thanks!
